src/validation/mod.rs

```diff
@@ -162,94 +164,96 @@ pub fn validate(wasm: &[u8]) -> Result<ValidationInfo> {
+    let elements: Vec<ElemType> = handle_section(&mut wasm, &mut header, SectionTy::Element, |wasm, _| {
+        use crate::core::reader::types::element::*;
+        use crate::RefType;
+        let mut elem_vec: Vec<ElemType> = Vec::new();
+        // https://webassembly.github.io/spec/core/binary/modules.html#element-section
+        // TODO: replace with wasm.read_vec in the future
+        let vec_length = wasm.read_var_u32().unwrap();
+        trace!("Element sections no.: {}", vec_length);
+        for i in 0..vec_length {
+            let ty = wasm.read_var_u32().unwrap();
+            // https://webassembly.github.io/spec/core/syntax/modules.html#element-segments
+            // https://webassembly.github.io/spec/core/binary/modules.html#element-section
+            // We can treat the ttype as a 3bit integer
+            // If it's not 3 bits I am not sure what to do
+            // bit 0 => diff between passive|declartive and active segment
+            // bit 1 => presence of an explicit table index for an active segment
+            // bit 2 => use of element type and element expressions instead of element kind and element indices
+            assert!((ty & 0b111) <= 0b111, "Element section is not encoded correctly. The type of this element is over 7 (0b111)");
+            // decide if we should
+            let elem_mode = if ty & 0b001 == 0b001 {
+                if ty & 0b010 == 0b010 {
+                    ElemMode::Declarative
+                } else {
+                    ElemMode::Passive
+                }
+            } else {
+                let table_idx = if ty & 0b010 == 0b010 {
+                    wasm.read_var_u32()?
+                } else {
+                    0
+                };
+                if tables.len() <= table_idx as usize {
+                    return Err(Error::UnknownTable);
+                }
+                let expr = read_constant_expression::read_constant_instructions(wasm, None, None)?;
+
+                ElemMode::Active(ActiveElem {
+                    table: table_idx,
+                    offset: expr,
+                })
+            };
+            let use_of_el_ty_and_el_exprs = ty & 0b100 == 0b100;
+
+            let reftype_or_elemkind: Option<RefType> = match if ty & 0b011 != 0 {
+                if use_of_el_ty_and_el_exprs {
+                    Some(wasm.read_u8()?)
+                } else {
+                    let read = wasm.read_u8()?;
+                    match read {
+                        0x00 => None,
+                        _ => panic!("Only FuncRefs are allowed"),
+                    }
+                }
+            } else {
+                None
+            } {
+                None => None,
+                Some(ty) => Some(RefType::from_byte(ty)?),
+            };
+
+            match reftype_or_elemkind {
+                Some(rty) => trace!("REFTYPE: {}", rty),
+                None => {trace!("REFTYPE NONE!")}
+            };
```

Comment:
I 100% know you are aware of this, but just leaving a tracking comment here to replace the panic with an error, and to pretty up the code (e.g. move the if outside  and/or chaining)